# Repaso de python y librerías específicas de Machine Learning


# Numpy
## Arrays

In [98]:
import numpy as np
arr = np.array([1,3,4,5,6]) # operar con matrices, vectores, tensores 
arr

array([1, 3, 4, 5, 6])

In [99]:
arr.shape

(5,)

In [100]:
arr.dtype

dtype('int32')

In [101]:
arr1 = np.array([1,'st','er',3]) # mezclo los tipos de datos
arr1.dtype

dtype('<U11')

El array tiene tipos de datos mezclados, por lo tanto se almacena como texto

In [102]:
arr1 #aca no puedo sumar 

array(['1', 'st', 'er', '3'], dtype='<U11')

In [103]:
np.sum(arr)

19

### Creando arrays

In [104]:
arr = np.array([[1,2,3],[2,4,6],[8,8,8]]) #matriz
arr.shape

(3, 3)

In [105]:
arr

array([[1, 2, 3],
       [2, 4, 6],
       [8, 8, 8]])

In [106]:
arr = np.zeros((2,4))
arr

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [107]:
arr = np.ones((2,4))
arr

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [108]:
arr = np.identity(3) #unos en la diagonal y es cuadrada de grado 3 
arr

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [109]:
arr = np.random.randn(3,4) #numeros aleatorios 
arr

array([[-0.23238983,  0.28796793, -0.2410245 , -0.45593829],
       [ 0.84304998,  0.51605027, -0.22895998, -0.57777928],
       [-0.48040871,  0.00437056, -0.06064173,  0.72026096]])

In [110]:
from io import BytesIO # representacion en texto y lo paso a numero 
b = BytesIO(b"2,23,33\n32,42,63.4\n35,77,12")
arr = np.genfromtxt(b, delimiter=",")
arr

array([[ 2. , 23. , 33. ],
       [32. , 42. , 63.4],
       [35. , 77. , 12. ]])

### Seleccionando elementos de un array
#### Indexado simple

In [111]:
arr[1] #accedo por el numero de elemento, empieza en 0 

array([32. , 42. , 63.4])

In [112]:
arr = np.arange(12).reshape(2,2,3) #empieza en 0 hasta 11, tiene 12 elementos y realiza el reshape, 3 columnas y 2 filas de c/matriz
arr

array([[[ 0,  1,  2],
        [ 3,  4,  5]],

       [[ 6,  7,  8],
        [ 9, 10, 11]]])

In [113]:
arr[0] #el primer elemento ahora es una matriz

array([[0, 1, 2],
       [3, 4, 5]])

In [114]:
arr = np.arange(10) #creo otro array de 10 elementos 
arr[5:] #desde el quinto elemento en adelante


array([5, 6, 7, 8, 9])

In [115]:
arr[5:8]

array([5, 6, 7])

In [116]:
arr[:-5]

array([0, 1, 2, 3, 4])

In [117]:
arr = np.arange(12).reshape(2,2,3)
arr

array([[[ 0,  1,  2],
        [ 3,  4,  5]],

       [[ 6,  7,  8],
        [ 9, 10, 11]]])

In [118]:
arr[1:2]

array([[[ 6,  7,  8],
        [ 9, 10, 11]]])

In [119]:
arr = np.arange(27).reshape(3,3,3) # 3 matrices de 3x3, 3 dimensiones
arr

array([[[ 0,  1,  2],
        [ 3,  4,  5],
        [ 6,  7,  8]],

       [[ 9, 10, 11],
        [12, 13, 14],
        [15, 16, 17]],

       [[18, 19, 20],
        [21, 22, 23],
        [24, 25, 26]]])

In [120]:
arr[:,:,2] # : (dos puntos) en las primeras dos posiciones significa que seleccionas todas las "capas"
           # El resultado es un array 2D con las terceras columnas de cada una de las capas (matriz)

array([[ 2,  5,  8],
       [11, 14, 17],
       [20, 23, 26]])

In [121]:
arr[...,2]

array([[ 2,  5,  8],
       [11, 14, 17],
       [20, 23, 26]])

#### Indexado advanzado

In [122]:
arr = np.arange(9).reshape(3,3)
arr

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

Esta selección está utilizando indexación avanzada en NumPy, lo que significa que está seleccionando elementos específicos del array usando dos listas de índices.

[0, 1, 2] son los índices de las filas.
[1, 0, 0] son los índices de las columnas.

Lo que NumPy hará es combinar cada índice de la primera lista (filas) con el índice correspondiente de la segunda lista (columnas) para seleccionar los elementos:

- El primer par de índices es (0, 1): Selecciona el elemento en la fila 0, columna 1, que es 1.

- El segundo par de índices es (1, 0): Selecciona el elemento en la fila 1, columna 0, que es 3.

- El tercer par de índices es (2, 0): Selecciona el elemento en la fila 2, columna 0, que es 6.

In [123]:
arr[[0,1,2],[1,0,0]]

array([1, 3, 6])

##### Indexado booleano

In [124]:
cities = np.array(["delhi","banglaore","mumbai","chennai","bhopal"])
city_data = np.random.randn(5,3)
city_data

array([[ 0.30650172, -0.17810627,  0.29500615],
       [ 1.85305745, -1.01535223, -0.04902214],
       [-0.37794336, -0.16536411,  2.1036866 ],
       [-0.97724974, -0.36901836, -0.72737483],
       [-2.71763277,  1.68548993,  0.20167489]])

In [125]:
city_data[cities =="delhi"]

array([[ 0.30650172, -0.17810627,  0.29500615]])

In [126]:
city_data[0]

array([ 0.30650172, -0.17810627,  0.29500615])

In [127]:
city_data[city_data > 0]

array([0.30650172, 0.29500615, 1.85305745, 2.1036866 , 1.68548993,
       0.20167489])

In [128]:
city_data[city_data > 0] = 0 # solo numeros neg
city_data


array([[ 0.        , -0.17810627,  0.        ],
       [ 0.        , -1.01535223, -0.04902214],
       [-0.37794336, -0.16536411,  0.        ],
       [-0.97724974, -0.36901836, -0.72737483],
       [-2.71763277,  0.        ,  0.        ]])

#### Operaciones sobre arrays

In [129]:
arr = np.arange(15).reshape(3,5)
arr

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

In [130]:
arr + 5 #sumo 5 a c/u de los elementos del array

array([[ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19]])

In [131]:
arr * 2 #multiplico por un escalar 

array([[ 0,  2,  4,  6,  8],
       [10, 12, 14, 16, 18],
       [20, 22, 24, 26, 28]])

In [132]:
arr1 = np.arange(15).reshape(5,3)
arr2 = np.arange(5).reshape(5,1)



In [133]:
arr1

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11],
       [12, 13, 14]])

In [134]:
arr2

array([[0],
       [1],
       [2],
       [3],
       [4]])

In [135]:
arr2 + arr1

array([[ 0,  1,  2],
       [ 4,  5,  6],
       [ 8,  9, 10],
       [12, 13, 14],
       [16, 17, 18]])

In [136]:
arr1 = np.random.randn(5,3)
arr1

array([[ 0.81635814, -0.32205837,  0.87227169],
       [ 0.24979975, -0.05700051,  0.0984454 ],
       [ 0.25123767,  0.615566  ,  0.15152085],
       [-0.05837062,  0.95558278, -0.50515131],
       [-0.01976601, -0.03667921,  1.40386295]])

In [137]:
np.modf(arr1) # Separa las partes entera y decimal

(array([[ 0.81635814, -0.32205837,  0.87227169],
        [ 0.24979975, -0.05700051,  0.0984454 ],
        [ 0.25123767,  0.615566  ,  0.15152085],
        [-0.05837062,  0.95558278, -0.50515131],
        [-0.01976601, -0.03667921,  0.40386295]]),
 array([[ 0., -0.,  0.],
        [ 0., -0.,  0.],
        [ 0.,  0.,  0.],
        [-0.,  0., -0.],
        [-0., -0.,  1.]]))

#### Algebra lineal usando numpy

In [138]:
A = np.array([[1,2,3],[4,5,6],[7,8,9]])
B = np.array([[9,8,7],[6,5,4],[1,2,3]])
A.dot(B) # producto matricial entre A  y B 


array([[ 24,  24,  24],
       [ 72,  69,  66],
       [120, 114, 108]])

C es una nueva matriz resultante, y cada elemento C[i,j] de la matriz resultante se calcula como el producto punto entre la fila 
i de la matriz 𝐴 y la columna 𝑗 de la matriz 𝐵.
El producto punto es una operación entre dos vectores que da como resultado un número (escalar) al multiplicar sus componentes correspondientes y sumar los productos.

El producto matricial consiste en calcular el producto punto entre las filas de 𝐴 y las columnas de 𝐵 generando una nueva matriz 𝐶 como resultado

In [139]:
A = np.arange(15).reshape(3,5)
A

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

In [140]:
A.T #transpuesta, cambia filas por columnas

array([[ 0,  5, 10],
       [ 1,  6, 11],
       [ 2,  7, 12],
       [ 3,  8, 13],
       [ 4,  9, 14]])

Resolver el sistema de ecuaciones x0 + 2 * x1 = 1 and 3 * x0 + 5 * x1 = 2:


In [141]:
a = np.array([[1, 2], [3, 5]])

b = np.array([1, 2])

x = np.linalg.solve(a, b)

x

array([-1.,  1.])

In [142]:
np.allclose(np.dot(a, x), b) #chequea la solución

True

# Pandas
## Data frames

In [143]:
import pandas as pd
d =  [{'city':'Delhi',"data":1000},  #diccionario
      {'city':'Banglaore',"data":2000},
      {'city':'Mumbai',"data":1000}]
pd.DataFrame(d)

city  data
0      Delhi  1000
1  Banglaore  2000
2     Mumbai  1000

In [144]:
df = pd.DataFrame(d)

### Leyendo datos

In [145]:
#from google.colab import files
#data_to_load = files.upload()

In [146]:
import io
import pandas as pd


city_data = pd.read_csv('2 - MLII - Clase 3 - datos.csv')

In [147]:
city_data.head(n=10)

city      city_ascii        lat        lng       pop  \
0   Qal eh-ye Now       Qal eh-ye  34.983000  63.133300    2997.0   
1     Chaghcharan     Chaghcharan  34.516701  65.250001   15000.0   
2     Lashkar Gah     Lashkar Gah  31.582998  64.360000  201546.0   
3          Zaranj          Zaranj  31.112001  61.886998   49851.0   
4      Tarin Kowt      Tarin Kowt  32.633298  65.866699   10000.0   
5    Zareh Sharan    Zareh Sharan  32.850000  68.416705   13737.0   
6        Asadabad        Asadabad  34.866000  71.150005   48400.0   
7         Taloqan         Taloqan  36.729999  69.540004   64256.0   
8  Mahmud-E Eraqi  Mahmud-E Eraqi  35.016696  69.333301    7407.0   
9      Mehtar Lam      Mehtar Lam  34.650000  70.166701   17345.0   

       country iso2 iso3 province  
0  Afghanistan   AF  AFG  Badghis  
1  Afghanistan   AF  AFG     Ghor  
2  Afghanistan   AF  AFG  Hilmand  
3  Afghanistan   AF  AFG   Nimroz  
4  Afghanistan   AF  AFG  Uruzgan  
5  Afghanistan   AF  AFG  Paktika  
6  Afghanistan   AF  AFG    Kunar  
7  Afghanistan   AF  AFG   Takhar  
8  Afghanistan   AF  AFG   Kapisa  
9  Afghanistan   AF  AFG  Laghman

In [148]:
city_data.tail()

city   city_ascii        lat        lng        pop   country  \
7317       Mutare       Mutare -18.970019  32.650038   216785.0  Zimbabwe   
7318       Kadoma       Kadoma -18.330006  29.909947    56400.0  Zimbabwe   
7319  Chitungwiza  Chitungwiza -18.000001  31.100003   331071.0  Zimbabwe   
7320       Harare       Harare -17.817790  31.044709  1557406.5  Zimbabwe   
7321     Bulawayo     Bulawayo -20.169998  28.580002   697096.0  Zimbabwe   

     iso2 iso3          province  
7317   ZW  ZWE        Manicaland  
7318   ZW  ZWE  Mashonaland West  
7319   ZW  ZWE            Harare  
7320   ZW  ZWE            Harare  
7321   ZW  ZWE          Bulawayo

In [149]:
series_es = city_data.lat # solo la columna latitud 

In [150]:
type(series_es)

pandas.core.series.Series

In [151]:
series_es[1:10:2] #va de a dos

1    34.516701
3    31.112001
5    32.850000
7    36.729999
9    34.650000
Name: lat, dtype: float64

In [152]:
series_es[:7] 

0    34.983000
1    34.516701
2    31.582998
3    31.112001
4    32.633298
5    32.850000
6    34.866000
Name: lat, dtype: float64

In [153]:
series_es[:-7315]

0    34.983000
1    34.516701
2    31.582998
3    31.112001
4    32.633298
5    32.850000
6    34.866000
Name: lat, dtype: float64

In [154]:
city_data[:7] #todas las columnas hasta el 7mo elemento

city    city_ascii        lat        lng       pop      country  \
0  Qal eh-ye Now     Qal eh-ye  34.983000  63.133300    2997.0  Afghanistan   
1    Chaghcharan   Chaghcharan  34.516701  65.250001   15000.0  Afghanistan   
2    Lashkar Gah   Lashkar Gah  31.582998  64.360000  201546.0  Afghanistan   
3         Zaranj        Zaranj  31.112001  61.886998   49851.0  Afghanistan   
4     Tarin Kowt    Tarin Kowt  32.633298  65.866699   10000.0  Afghanistan   
5   Zareh Sharan  Zareh Sharan  32.850000  68.416705   13737.0  Afghanistan   
6       Asadabad      Asadabad  34.866000  71.150005   48400.0  Afghanistan   

  iso2 iso3 province  
0   AF  AFG  Badghis  
1   AF  AFG     Ghor  
2   AF  AFG  Hilmand  
3   AF  AFG   Nimroz  
4   AF  AFG  Uruzgan  
5   AF  AFG  Paktika  
6   AF  AFG    Kunar

In [155]:
city_data.iloc[:5,:4] # 5 filas y las 4 primeras columnas

city   city_ascii        lat        lng
0  Qal eh-ye Now    Qal eh-ye  34.983000  63.133300
1    Chaghcharan  Chaghcharan  34.516701  65.250001
2    Lashkar Gah  Lashkar Gah  31.582998  64.360000
3         Zaranj       Zaranj  31.112001  61.886998
4     Tarin Kowt   Tarin Kowt  32.633298  65.866699

In [156]:
city_data[city_data['pop'] > 10000000][city_data.columns[pd.Series(city_data.columns).str.startswith('l')]]
# selecciono las columnas que empiezan con l donde la poblacion sea mayor a 10000000 habitantes 

lat         lng
360  -34.602502  -58.397531
1171 -23.558680  -46.625020
2068  31.216452  121.436505
3098  28.669993   77.230004
3110  19.016990   72.856989
3492  35.685017  139.751407
4074  19.442442  -99.130988
4513  24.869992   66.990009
5394  55.752164   37.615523
6124  41.104996   29.010002
7071  40.749979  -73.980017

In [157]:
city_greater_10mil = city_data[city_data['pop'] > 10000000]
city_greater_10mil

city    city_ascii        lat         lng         pop  \
360   Buenos Aires  Buenos Aires -34.602502  -58.397531  11862073.0   
1171     Sao Paulo     Sao Paulo -23.558680  -46.625020  14433147.5   
2068      Shanghai      Shanghai  31.216452  121.436505  14797756.0   
3098         Delhi         Delhi  28.669993   77.230004  11779606.5   
3110        Mumbai        Mumbai  19.016990   72.856989  15834918.0   
3492         Tokyo         Tokyo  35.685017  139.751407  22006299.5   
4074   Mexico City   Mexico City  19.442442  -99.130988  14919501.0   
4513       Karachi       Karachi  24.869992   66.990009  11877109.5   
5394        Moscow        Moscow  55.752164   37.615523  10452000.0   
6124      Istanbul      Istanbul  41.104996   29.010002  10003305.0   
7071      New York      New York  40.749979  -73.980017  13524139.0   

                       country iso2 iso3                province  
360                  Argentina   AR  ARG  Ciudad de Buenos Aires  
1171                    Brazil   BR  BRA               São Paulo  
2068                     China   CN  CHN                Shanghai  
3098                     India   IN  IND                   Delhi  
3110                     India   IN  IND             Maharashtra  
3492                     Japan   JP  JPN                   Tokyo  
4074                    Mexico   MX  MEX        Distrito Federal  
4513                  Pakistan   PK  PAK                    Sind  
5394                    Russia   RU  RUS                  Moskva  
6124                    Turkey   TR  TUR                Istanbul  
7071  United States of America   US  USA                New York

In [158]:
city_greater_10mil.rename(columns={'pop':'population'}, inplace=True)
# con implace modifico el dataset original

C:\Users\carlos\AppData\Local\Temp\ipykernel_22420\1069009551.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  city_greater_10mil.rename(columns={'pop':'population'}, inplace=True)


In [159]:
city_greater_10mil.where(city_greater_10mil.population > 15000000)
#cuando no es mayor a 15M me devuelve NAN

city city_ascii        lat         lng  population country iso2 iso3  \
360      NaN        NaN        NaN         NaN         NaN     NaN  NaN  NaN   
1171     NaN        NaN        NaN         NaN         NaN     NaN  NaN  NaN   
2068     NaN        NaN        NaN         NaN         NaN     NaN  NaN  NaN   
3098     NaN        NaN        NaN         NaN         NaN     NaN  NaN  NaN   
3110  Mumbai     Mumbai  19.016990   72.856989  15834918.0   India   IN  IND   
3492   Tokyo      Tokyo  35.685017  139.751407  22006299.5   Japan   JP  JPN   
4074     NaN        NaN        NaN         NaN         NaN     NaN  NaN  NaN   
4513     NaN        NaN        NaN         NaN         NaN     NaN  NaN  NaN   
5394     NaN        NaN        NaN         NaN         NaN     NaN  NaN  NaN   
6124     NaN        NaN        NaN         NaN         NaN     NaN  NaN  NaN   
7071     NaN        NaN        NaN         NaN         NaN     NaN  NaN  NaN   

         province  
360           NaN  
1171          NaN  
2068          NaN  
3098          NaN  
3110  Maharashtra  
3492        Tokyo  
4074          NaN  
4513          NaN  
5394          NaN  
6124          NaN  
7071          NaN

In [160]:
city_greater_10mil[city_greater_10mil.population > 15000000] # solo me devuelve las que me interesan

city city_ascii        lat         lng  population country iso2 iso3  \
3110  Mumbai     Mumbai  19.016990   72.856989  15834918.0   India   IN  IND   
3492   Tokyo      Tokyo  35.685017  139.751407  22006299.5   Japan   JP  JPN   

         province  
3110  Maharashtra  
3492        Tokyo

### Operaciones sobre dataframes

In [161]:
df = pd.DataFrame(np.random.randn(8, 3),
columns=['A', 'B', 'C'])

In [162]:
nparray = df.values # los valores del dataframe los puedo convertir en un array, hacer operaciones y volverlo a meter en un dataframe
type(nparray)

numpy.ndarray

In [163]:
df

A         B         C
0 -0.693670  0.236564  2.062386
1 -0.004486  0.869050 -1.837513
2 -1.573256 -0.024910  0.651718
3  2.106830  0.588231  0.513573
4  0.398093 -1.297165  0.041006
5 -0.281750 -0.312493  1.139094
6  0.324952  0.361828  0.592348
7  0.297024  0.196043 -0.064191

In [164]:
from numpy import nan
df.iloc[4,2] = nan  # reemplazo por nan

In [165]:
df

A         B         C
0 -0.693670  0.236564  2.062386
1 -0.004486  0.869050 -1.837513
2 -1.573256 -0.024910  0.651718
3  2.106830  0.588231  0.513573
4  0.398093 -1.297165       NaN
5 -0.281750 -0.312493  1.139094
6  0.324952  0.361828  0.592348
7  0.297024  0.196043 -0.064191

In [166]:
df.fillna(0)

A         B         C
0 -0.693670  0.236564  2.062386
1 -0.004486  0.869050 -1.837513
2 -1.573256 -0.024910  0.651718
3  2.106830  0.588231  0.513573
4  0.398093 -1.297165  0.000000
5 -0.281750 -0.312493  1.139094
6  0.324952  0.361828  0.592348
7  0.297024  0.196043 -0.064191

In [167]:
columns_numeric = ['lat','lng','pop']

In [168]:
city_data[columns_numeric].mean()

lat        20.662876
lng        10.711914
pop    265463.071633
dtype: float64

In [169]:
city_data[columns_numeric].sum()

lat    1.512936e+05
lng    7.843263e+04
pop    1.943721e+09
dtype: float64

In [170]:
city_data[columns_numeric].count()

lat    7322
lng    7322
pop    7322
dtype: int64

In [171]:
city_data[columns_numeric].median()

lat       26.792730
lng       18.617509
pop    61322.750000
dtype: float64

In [172]:
city_data[columns_numeric].quantile(0.8)

lat        46.852480
lng        89.900018
pop    269210.000000
Name: 0.8, dtype: float64

In [173]:
city_data[columns_numeric].sum(axis = 1).head()

0      3095.116300
1     15099.766702
2    201641.942998
3     49943.998999
4     10098.499997
dtype: float64

In [174]:
city_data[columns_numeric].describe()

lat          lng           pop
count  7322.000000  7322.000000  7.322000e+03
mean     20.662876    10.711914  2.654631e+05
std      29.134818    79.044615  8.287622e+05
min     -89.982894  -179.589979 -9.900000e+01
25%      -0.324710   -64.788472  1.734425e+04
50%      26.792730    18.617509  6.132275e+04
75%      43.575448    73.103628  2.001726e+05
max      82.483323   179.383304  2.200630e+07

In [175]:
city_data1 = city_data.sample(3)
city_data1

city              city_ascii        lat         lng  \
6493              Sand Point              Sand Point  55.339709 -160.497191   
5742               Harnosand               Harnosand  62.633997   17.934004   
260   Concepcion del Uruguay  Concepcion del Uruguay -32.479996  -58.239996   

          pop                   country iso2 iso3        province  
6493    667.0  United States of America   US  USA          Alaska  
5742  17016.0                    Sweden   SE  SWE  Västernorrland  
260   48275.0                 Argentina   AR  ARG      Entre Ríos

### Concatenando data frames

In [176]:
city_data2 = city_data.sample(3)
city_data2

city   city_ascii        lat        lng       pop  \
3256     Chabahar     Chabahar  25.300405  60.629932   56544.0   
6641  St. Charles  St. Charles  38.784285 -90.506166  213139.5   
2817  Kissidougou  Kissidougou   9.190458 -10.120013   56957.0   

                       country iso2 iso3                province  
3256                      Iran   IR  IRN  Sistan and Baluchestan  
6641  United States of America   US  USA                Missouri  
2817                    Guinea   GN  GIN                 Faranah

In [177]:
city_data1

city              city_ascii        lat         lng  \
6493              Sand Point              Sand Point  55.339709 -160.497191   
5742               Harnosand               Harnosand  62.633997   17.934004   
260   Concepcion del Uruguay  Concepcion del Uruguay -32.479996  -58.239996   

          pop                   country iso2 iso3        province  
6493    667.0  United States of America   US  USA          Alaska  
5742  17016.0                    Sweden   SE  SWE  Västernorrland  
260   48275.0                 Argentina   AR  ARG      Entre Ríos

In [178]:
city_data_combine = pd.concat([city_data1,city_data2])
city_data_combine

# axis{0/’index’, 1/’columns’}, default 0

city              city_ascii        lat         lng  \
6493              Sand Point              Sand Point  55.339709 -160.497191   
5742               Harnosand               Harnosand  62.633997   17.934004   
260   Concepcion del Uruguay  Concepcion del Uruguay -32.479996  -58.239996   
3256                Chabahar                Chabahar  25.300405   60.629932   
6641             St. Charles             St. Charles  38.784285  -90.506166   
2817             Kissidougou             Kissidougou   9.190458  -10.120013   

           pop                   country iso2 iso3                province  
6493     667.0  United States of America   US  USA                  Alaska  
5742   17016.0                    Sweden   SE  SWE          Västernorrland  
260    48275.0                 Argentina   AR  ARG              Entre Ríos  
3256   56544.0                      Iran   IR  IRN  Sistan and Baluchestan  
6641  213139.5  United States of America   US  USA                Missouri  
2817   56957.0                    Guinea   GN  GIN                 Faranah

In [179]:
df1 = pd.DataFrame({'col1': ['col10', 'col11', 'col12', 'col13'],
                    'col2': ['col20', 'col21', 'col22', 'col23'],
                    'col3': ['col30', 'col31', 'col32', 'col33'],
                    'col4': ['col40', 'col41', 'col42', 'col43']},
                   index=[0, 1, 2, 3])

In [180]:
df1

col1   col2   col3   col4
0  col10  col20  col30  col40
1  col11  col21  col31  col41
2  col12  col22  col32  col42
3  col13  col23  col33  col43

In [181]:
df4 = pd.DataFrame({'col2': ['col22', 'col23', 'col26', 'col27'],
                    'Col4': ['Col42', 'Col43', 'Col46', 'Col47'],
                    'col6': ['col62', 'col63', 'col66', 'col67']},
                   index=[2, 3, 6, 7])

df4

col2   Col4   col6
2  col22  Col42  col62
3  col23  Col43  col63
6  col26  Col46  col66
7  col27  Col47  col67

In [182]:
pd.concat([df1,df4], axis=1) # agrega las columnas, donde no puede concatenar pone un NaN 

col1   col2   col3   col4   col2   Col4   col6
0  col10  col20  col30  col40    NaN    NaN    NaN
1  col11  col21  col31  col41    NaN    NaN    NaN
2  col12  col22  col32  col42  col22  Col42  col62
3  col13  col23  col33  col43  col23  Col43  col63
6    NaN    NaN    NaN    NaN  col26  Col46  col66
7    NaN    NaN    NaN    NaN  col27  Col47  col67

In [183]:
country_data = city_data[['iso3','country']].drop_duplicates()

In [184]:
country_data.shape

(223, 2)

In [185]:
country_data.head()

iso3         country
0    AFG     Afghanistan
33   ALD           Aland
34   ALB         Albania
60   DZA         Algeria
111  ASM  American Samoa

In [186]:
del(city_data['country']) # borro la columna pais en el original

In [187]:
city_data.merge(country_data, 'inner').head() # tengo una tabla con la letra pais y otra con el nombre del pais, la PK es iso3

city   city_ascii        lat        lng       pop iso2 iso3  \
0  Qal eh-ye Now    Qal eh-ye  34.983000  63.133300    2997.0   AF  AFG   
1    Chaghcharan  Chaghcharan  34.516701  65.250001   15000.0   AF  AFG   
2    Lashkar Gah  Lashkar Gah  31.582998  64.360000  201546.0   AF  AFG   
3         Zaranj       Zaranj  31.112001  61.886998   49851.0   AF  AFG   
4     Tarin Kowt   Tarin Kowt  32.633298  65.866699   10000.0   AF  AFG   

  province      country  
0  Badghis  Afghanistan  
1     Ghor  Afghanistan  
2  Hilmand  Afghanistan  
3   Nimroz  Afghanistan  
4  Uruzgan  Afghanistan

# Scikit-learn

## Precios de casas en Boston 

Detalles:

    Number of Instances 506
    Number of Attributes 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.
    
    Attribute Information (in order)

            CRIM per capita crime rate by town

            ZN proportion of residential land zoned for lots over 25,000 sq.ft.

            INDUS proportion of non-retail business acres per town

            CHAS Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)

            NOX nitric oxides concentration (parts per 10 million)

            RM average number of rooms per dwelling

            AGE proportion of owner-occupied units built prior to 1940

            DIS weighted distances to five Boston employment centres

            RAD index of accessibility to radial highways

            TAX full-value property-tax rate per $10,000

            PTRATIO pupil-teacher ratio by town

            B 1000(Bk - 0.63)^2 where Bk is the proportion of black people by town

            LSTAT % lower status of the population

            MEDV Median value of owner-occupied homes in $1000’s

    Missing Attribute Values

        None
    Creator

        Harrison, D. and Rubinfeld, D.L.

This is a copy of UCI ML housing dataset. https://archive.ics.uci.edu/ml/machine-learning-databases/housing/

This dataset was taken from the StatLib library which is maintained at Carnegie Mellon University.

The Boston house-price data of Harrison, D. and Rubinfeld, D.L. ‘Hedonic prices and the demand for clean air’, J. Environ. Economics & Management, vol.5, 81-102, 1978. Used in Belsley, Kuh & Welsch, ‘Regression diagnostics …’, Wiley, 1980. N.B. Various transformations are used in the table on pages 244-261 of the latter.

The Boston house-price data has been used in many machine learning papers that address regression problems.


In [188]:
from sklearn import datasets
boston = datasets.load_boston()


ImportError: 
`load_boston` has been removed from scikit-learn since version 1.2.

The Boston housing prices dataset has an ethical problem: as
investigated in [1], the authors of this dataset engineered a
non-invertible variable "B" assuming that racial self-segregation had a
positive impact on house prices [2]. Furthermore the goal of the
research that led to the creation of this dataset was to study the
impact of air quality but it did not give adequate demonstration of the
validity of this assumption.

The scikit-learn maintainers therefore strongly discourage the use of
this dataset unless the purpose of the code is to study and educate
about ethical issues in data science and machine learning.

In this special case, you can fetch the dataset from the original
source::

    import pandas as pd
    import numpy as np

    data_url = "http://lib.stat.cmu.edu/datasets/boston"
    raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
    data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
    target = raw_df.values[1::2, 2]

Alternative datasets include the California housing dataset and the
Ames housing dataset. You can load the datasets as follows::

    from sklearn.datasets import fetch_california_housing
    housing = fetch_california_housing()

for the California housing dataset and::

    from sklearn.datasets import fetch_openml
    housing = fetch_openml(name="house_prices", as_frame=True)

for the Ames housing dataset.

[1] M Carlisle.
"Racist data destruction?"
<https://medium.com/@docintangible/racist-data-destruction-113e3eff54a8>

[2] Harrison Jr, David, and Daniel L. Rubinfeld.
"Hedonic housing prices and the demand for clean air."
Journal of environmental economics and management 5.1 (1978): 81-102.
<https://www.researchgate.net/publication/4974606_Hedonic_housing_prices_and_the_demand_for_clean_air>


In [121]:
from sklearn.datasets import fetch_california_housing
boston = fetch_california_housing()

In [ ]:
boston

In [124]:
X = boston.data
y = boston.target

In [ ]:
X[:5]

In [ ]:
y[:10]

In [127]:
feature_names=['CRIM','ZN','INDUS','CHAS','NOX','RM','AGE','DIS','RAD','TAX','PTRATIO','B','LSTAT']

## Regresión

Regresión Lineal:

En la regresión lineal, tratamos de ajustar una línea recta que relacione la variable dependiente 𝑦 con las variables independientes 
X. El modelo predice 𝑦 a partir de 𝑋 minimizando la suma de los errores.

La regularización es una técnica que introduce una penalización en el modelo para evitar el sobreajuste y reducir la complejidad del modelo.

https://github.com/bcamandone/Machine-Learning/blob/main/Documentos%20te%C3%B3ricos/Sesgo%20y%20Varianza.pdf

### ¿Cómo funciona Lasso?

- Penalización L1: Lasso introduce la penalización en la forma de la norma L1 de los coeficientes de las variables independientes. Esto fuerza a algunos coeficientes a hacerse exactamente cero, lo que tiene el efecto de seleccionar características importantes y descartar las menos relevantes.
    
    
- Selección de variables: Una de las propiedades más importantes de Lasso es que realiza una selección automática de variables. Las características con coeficientes que se vuelven cero son efectivamente eliminadas del modelo. Esto simplifica el modelo al enfocarse solo en las características que realmente importan.
    
- Reducir el overfitting: Al penalizar las variables irrelevantes y reducir los coeficientes grandes, Lasso ayuda a crear un modelo más simple y generalizable.

In [128]:
from sklearn import datasets
from sklearn.linear_model import Lasso

from sklearn import linear_model, datasets
from sklearn.model_selection import GridSearchCV



GridSearchCV es una técnica de búsqueda exhaustiva utilizada en Machine Learning para encontrar los mejores hiperparámetros para un modelo. Es especialmente útil cuando estás trabajando con modelos que tienen hiperparámetros que afectan su rendimiento, y deseas encontrar la combinación que optimice los resultados.

En Machine Learning, los hiperparámetros son parámetros que no se aprenden directamente del modelo, sino que deben ser establecidos antes del proceso de entrenamiento.

### ¿Cómo funciona GridSearchCV?

Especificar el modelo y los hiperparámetros:

Seleccionas el modelo que deseas ajustar (como un árbol de decisión, regresión logística, etc.).
Luego, defines un conjunto de hiperparámetros que deseas probar. Cada hiperparámetro tendrá un conjunto de valores posibles  (como un rango de valores de regularización, el número de vecinos para k-NN, etc.).

Búsqueda en la cuadrícula (grid):

GridSearchCV crea una "cuadrícula" de todas las combinaciones posibles de los hiperparámetros que has especificado.
Por ejemplo, si tienes dos hiperparámetros con 3 valores posibles cada uno, se probarán 9 combinaciones en total.
Entrenamiento y validación cruzada:

Para cada combinación de hiperparámetros, GridSearchCV entrena el modelo utilizando validación cruzada, dividendo el conjunto de datos en varias particiones (folds).

El modelo se entrena en una parte y se valida en la otra, repitiendo esto para todas las particiones.

Evaluación del rendimiento:

Después de probar todas las combinaciones, GridSearchCV evalúa el rendimiento del modelo (por ejemplo, utilizando precisión, F1-score, o cualquier métrica relevante) y selecciona la mejor combinación de hiperparámetros basada en el rendimiento promedio en las validaciones cruzadas.

Mejores hiperparámetros:

Una vez que se completa la búsqueda, GridSearchCV te dice cuál es la mejor combinación de hiperparámetros que maximiza el rendimiento del modelo.

https://scikit-learn.org/dev/api/sklearn.linear_model.html

In [132]:

X_train = boston.data[:310]  # primeras 310 para entrenar 
y_train = boston.target[:310]

X_test = boston.data[310:]
y_test = boston.target[310:]



In [ ]:
lasso = Lasso(random_state=0) # semilla para reproducir el mismo experimento 

alphas = np.logspace(-4, -0.5, 30) # parametro alpha por defecto vale 1 

scores = list() # lista para guardar los scores, diferencia entre valor del precio del inmueble y valos predicho
scores_std = list()

estimator = GridSearchCV(lasso,
                         param_grid = dict(alpha=alphas)) # prueba con todos los valores de alpha hasta que encuentra el mejor

estimator.fit(X_train, y_train)

In [ ]:
estimator.best_score_ #menor diferencia en promedio entre el valor predicho y el valor real R²

El atributo estimator.best_score_ te indica el mejor puntaje (score) obtenido durante la validación cruzada utilizando los hiperparámetros óptimos que fueron encontrados durante el proceso de búsqueda.

Este valor indica que el modelo captura parte de la relación entre las variables, pero, también sugiere que hay margen para mejorar el rendimiento, ya que el 59.2% de la variabilidad de los precios de las casas no se está explicando.

In [ ]:
estimator.best_estimator_

In [137]:
pred = estimator.predict(X_test) # uso el conjunto de test 

In [138]:
dif = pred - y_test # Diferencia entre la predicción y el valor real

In [ ]:
from matplotlib import pyplot as plt 
plt.hist(dif, bins = 50) 
plt.title("Diferencia entre la predicción y el valor real") #histograma de la diferencia 
plt.show()

 Valores negativos: Mi predicción es mucho mas chica que el valor real 
 
 
 Debería tener un histograma centrado en 0, la diferencia es cercana a cero 

## Ejercicio: buscar otro modelo de regresión en la documentación de scikit learn y probarlo